// egg-freelog-base/app/extend/context.js
/**
 * Created by yuliang on 2017/6/30.
 */

module.exports = {
    /**
     * 定义成功返回的API数据结构
     * @param data
     * @returns {exports}
     */
    success(data){

        let {retCodeEnum, errCodeEnum} = this.app

        this.body = this.buildReturnObject(retCodeEnum.success, errCodeEnum.success, 'success', data)

        return this
    },

    /**
     * 只接受JSON格式的消息体
     * @param msg
     * @returns {exports}
     */
    allowContentType({type = 'json', msg = ''}){
        if (!this.is(type)) {
            throw new Error(msg || `content-type must be ${type}`)
        }
        return this
    },

    /**
     * 定义错误处理->此处抛出异常会由middleware=>error_handler处理
     * @param msg
     * @param errCode
     * @param retCode
     */
    error ({msg, errCode, retCode, data})  {
        let {retCodeEnum, errCodeEnum, type} = this.app

        let message = type.error(arguments[0])
            ? arguments[0].message || arguments[0].toString()
            : type.string(arguments[0]) ? arguments[0] : msg || '接口口内部异常'

        throw Object.assign(new Error(message), {
            retCode: retCode ? retCode : retCodeEnum.serverError,
            errCode: errCode ? errCode : errCodeEnum.apiError,
            data
        })
    },

    /**
     * 校验参数,有错误就抛出异常
     * @returns {exports}
     */
    validate(){
        if (!this.errors || this.errors.length < 1) {
            return this
        }
        let msg = '参数校验失败,details:' + JSON.stringify(this.errors);
        this.error({
            msg: msg,
            errCode: this.app.errCodeEnum.paramTypeError
        })
    },

    /**
     * 构建API返回数据格式
     * @param ret {int} 一级错误码
     * @param errCode {int} 二级错误码
     * @param msg  {string} 消息内容
     * @param data {object} 返回数据
     * @returns {{ret: number, errcode: number, msg: string}}
     */
    buildReturnObject(ret, errCode, msg, data){
        let {type, retCodeEnum, errCodeEnum} = this.app
        let result = {
            ret: type.int32(ret) ? ret : retCodeEnum.success,
            errcode: type.int32(errCode) ? errCode : errCodeEnum.success,
            msg: type.string(msg) || type.number(msg) ? msg.toString() : "success",
            data: type.nullOrUndefined(data) ? null : data
        }
        return result
    },

    /**
     * 获取内部REST-API数据
     * @param url
     * @param options
     */
    curlIntranetApi(url, options){
        let opt = Object.assign({
            headers: {'auth-token': this.headers['auth-token']},
            dataType: 'json'
        }, options)

        return this.curl(url, opt).then(data => {
            return data.data
        }).then(this.helper.convertApiResult)
    }
}
